### PR TITLE
KubernetesWatchTask Prerequisite refactor in JKube Kit

### PR DESCRIPTION
**Pull request by rohanKanojia**
## Description

Prerequisite Refactor in JKube Kit for #1109
+ SpringBootWatcher.isApplicable returns true for Spring Boot Gradle Plugin
+ Remove JKubeArtifactResolverService; Add utility method in JKubeProjectUtil to resolve artifact from dependencies instead
+ SpringBootUtil.getSpringBootPluginConfiguration also fetches
  configuration of Spring Boot Gradle Plugin

Signed-off-by: Rohan Kumar <rohaan@redhat.com>

## Type of change

 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [X] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [X] I have read the [contributing guidelines](https://www.eclipse.org/jkube/contributing)
 - [X] I signed-off my commit with a user that has signed the [Eclipse Contributor Agreement](https://www.eclipse.org/legal/ECA.php)
 - [ ] I Added [CHANGELOG](../CHANGELOG.md) entry
 - [X] I have implemented unit tests to cover my changes
 - [ ] I have updated the [documentation](../kubernetes-maven-plugin/doc) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=jkubeio_jkube) report
 - [ ] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

### CODE REVIEW

File: jkube-kit/common-maven/src/test/java/org/eclipse/jkube/kit/common/util/SpringBootUtilTest.java [hunk 1]
```diff
@@ -74,4 +81,42 @@ public void testMultipleProfilesParsing() {
         assertNull(props.get("spring.application.name"));
     }
 
+    @Test
+    public void getSpringBootPluginConfiguration_whenSpringBootMavenPluginPresent_thenReturnsPluginConfiguration() {
+        // Given
+        JavaProject javaProject = JavaProject.builder()
+            .plugin(Plugin.builder()
+                .groupId(SPRING_BOOT_GROUP_ID)
+                .artifactId(SPRING_BOOT_MAVEN_PLUGIN_ARTIFACT_ID)
```
Comment (manusa) [inline jkube-kit/common-maven/src/test/java/org/eclipse/jkube/kit/common/util/SpringBootUtilTest.java:90]: Please, use a string here, if there's an error/typo in the constant it won't be revealed

File: jkube-kit/common-maven/src/test/java/org/eclipse/jkube/kit/common/util/SpringBootUtilTest.java [hunk 2]
```diff
@@ -74,4 +81,42 @@ public void testMultipleProfilesParsing() {
...
+        assertEquals("ZIP", configuration.get("layout").toString());
+    }
+
+    @Test
+    public void getSpringBootPluginConfiguration_whenSpringBootGradlePluginPresent_thenReturnsPluginConfiguration() {
+        // Given
+        JavaProject javaProject = JavaProject.builder()
+            .plugin(Plugin.builder()
+                .groupId(SPRING_BOOT_GROUP_ID)
+                .artifactId(SPRING_BOOT_GRADLE_PLUGIN_ARTIFACT_ID)
```
Comment (manusa) [inline jkube-kit/common-maven/src/test/java/org/eclipse/jkube/kit/common/util/SpringBootUtilTest.java:109]: https://github.com/eclipse/jkube/pull/1127/files#r754970391

File: kubernetes-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/build/ApplyMojo.java [hunk 3]
```diff
@@ -42,7 +42,7 @@
 /**
  * Base class for goals which deploy the generated artifacts into the Kubernetes cluster
  */
-@Mojo(name = "apply", requiresDependencyResolution = ResolutionScope.COMPILE, defaultPhase = LifecyclePhase.INSTALL)
+@Mojo(name = "apply", requiresDependencyResolution = ResolutionScope.COMPILE_PLUS_RUNTIME, defaultPhase = LifecyclePhase.INSTALL)
```
Comment (manusa) [inline kubernetes-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/build/ApplyMojo.java:45]: Same change should be applicable to every child:
- OpenshiftApplyMojo
- DebugMojo
- OpenshiftDebugMojo
- DeployMojo
- OpenshiftDeployMojo
- LogMojo
- OpenshiftLogMojo

File: kubernetes-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/develop/WatchMojo.java [hunk 4]
```diff
@@ -60,7 +60,7 @@
 /**
  * Used to automatically rebuild Docker images and restart containers in case of updates.
  */
-@Mojo(name = "watch", defaultPhase = LifecyclePhase.PACKAGE, requiresDependencyResolution = ResolutionScope.COMPILE)
+@Mojo(name = "watch", defaultPhase = LifecyclePhase.PACKAGE, requiresDependencyResolution = ResolutionScope.COMPILE_PLUS_RUNTIME)
```
Comment (manusa) [inline kubernetes-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/develop/WatchMojo.java:63]: Same change should be applicable to every child:
- OpenshiftWatchMojo

File: jkube-kit/common/src/main/java/org/eclipse/jkube/kit/common/util/JKubeProjectUtil.java [hunk 5]
```diff
@@ -177,4 +177,19 @@ public static String getProperty(String key, JavaProject project) {
...
+            && d.getVersion().equals(version)
+            && d.getType().equals(type))
+        .findFirst()
+        .map(Dependency::getFile)
+        .orElse(null);
+    if (artifact == null) {
+      throw new IllegalStateException("Cannot find artifact " + String.format("%s-%s.%s", artifactId, version, type) + " within the resolved resources");
+    }
+    return artifact;
+  }
```
Comment (manusa) [inline jkube-kit/common/src/main/java/org/eclipse/jkube/kit/common/util/JKubeProjectUtil.java:194]: We already have a getDependency and getPlugin.
Do we really need another method for this?
If yes, is the exception thrown here appropriate or something we want?

File: kubernetes-maven-plugin/plugin/src/test/java/org/eclipse/jkube/maven/plugin/mojo/GeneratedPluginDescriptorTest.java [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+  public static Collection<Object[]> data() {
+    return Arrays.asList(
+        new Object[] { "build", "compile", "pre-integration-test"},
+        new Object[] { "resource", "compile", "process-resources"},
+        new Object[] { "apply", "compile+runtime", "install"},
+        new Object[] { "deploy", "compile", "validate"},
+        new Object[] { "watch", "compile+runtime", "package"},
+        new Object[] { "undeploy", "compile", "install"},
+        new Object[] { "debug", "compile", "package"},
+        new Object[] { "log", "compile", "validate"}
```
Comment (manusa) [inline kubernetes-maven-plugin/plugin/src/test/java/org/eclipse/jkube/maven/plugin/mojo/GeneratedPluginDescriptorTest.java:64]: I really like what you did here. However, we should add **all** of the mojos.

File: jkube-kit/common/src/main/java/org/eclipse/jkube/kit/common/util/JKubeProjectUtil.java (hunk 5 above)
Comment (rohanKanojia) [reply jkube-kit/common/src/main/java/org/eclipse/jkube/kit/common/util/JKubeProjectUtil.java:194]: It's used by two classes: SpringBootGenerator and SpringBootWatcher, I can move this code to these calling places if you want.

**Review by manusa — APPROVED** [review]
LGTM, thx!

**Pull request merged**